Some addition and fixes to the initial design (#1)
nrc/sig-transaction: design/parallel-commit/initial-design.md

### TiDB

* The user should opt-in to parallel commit; we assume the user has opted in for all nodes for the rest of this document.

> * Each prewrite response will include a `min_commit_ts`, TiDB will select the largest `min_commit_ts` and add one to make the `commit_ts` for the transaction.

Reviewer: ```suggestion
* Each prewrite response will include a `min_commit_ts`, TiDB will select the largest `min_commit_ts` as the final `commit_ts` for the transaction.
```

* TiDB can return success to the client before sending the commit message but after receiving success responses for all prewrite messages.

> * If an operation fails because it is locked, TiDB must query the primary lock to find a list of secondaries (a `CheckTxnStatus` request). It can then send a batch get request to each region to get all secondary locks and the `min_commit_ts` in the locks. If all locks are got, it can send the commit message as above. If any lock disappears, the primary key must have been committed or rolled back. Use `CheckTxnStatus` to get the updated information and resolve the locks.

Reviewer 2: How are rollbacks initiated?

Author: I think it's the same as now so it's not worth mentioning. If TTL exceeds, `CheckTxnStatus` rolls back the primary lock, otherwise it returns the secondary list.

  - A structure of `min_commit_ts`s, a map from each in-progress transaction to the minimum ts at which it may be committed.
  - `max_read_ts`: the largest `start_ts` for any transactional read operation for the region (i.e., this value is potentially set on every read).
  - When a TiKV node is started up or becomes leader, `max_read_ts` is initialised from PD with a new timestamp.
  - When a prewrite is processed, TiKV records the current `max_read_ts + 1` as the `min_commit_ts` for that transaction. `min_commit_ts` is recorded in each key's lock data structure.
  - When a prewrite is finished, its entry is removed from the `min_commit_ts` structure. If the prewrite is successful, the `min_commit_ts` is returned to TiDB in the response.

>   - When a read is processed, first it sets the `max_read_ts`, then it checks its `start_ts` against the smallest `min_commit_ts` of any current transaction in the read range. It will block until its `start_ts >= min(min_commit_ts)`

Reviewer 2: "in the read range" - I thought we would do this later and start by conservatively using the whole region?

Reviewer: I think we can do it in the beginning. The whole region way isn't much easier(I think) and we must adopt the key level way.
